### DISCUSSION TITLE
Installing Ubuntu fails with error `0x8007007e The specified module could not be found`.

### DISCUSSION THREAD

**Discussion by JesseTG** [microsoft/WSL#8359] [Q&A] [answered]
### Version

Microsoft Windows [Version 10.0.19044.1645]

### WSL Version

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

N/A (can't get this far)

### Distro Version

Ubuntu 22.04

### Other Software

None that are relevant

### Repro Steps

1. Enable WSL as defined in the [instructions](https://docs.microsoft.com/en-us/windows/wsl/install)
2. Reboot
3. Install Ubuntu 22.04 from the Windows store
4. Attempt to launch Ubuntu
5. See the error message

### Expected Behavior

Ubuntu should install and launch successfully.

### Actual Behavior

This:

```
Installing, this may take a few minutes...
WslRegisterDistribution failed with error: 0x8007007e
Error: 0x8007007e The specified module could not be found.

Press any key to continue...
```

### Diagnostic Logs

https://aka.ms/AAgrjjg

**Comment by ghost** [comment]
Unfortunately I don't see any logs associated with this feedback? Could you follow 
 this? https://github.com/Microsoft/WSL/blob/master/CONTRIBUTING.md#record-wsl-logs-manually

**Comment by JesseTG** [comment]
@pmartincic I already did, and I thought they'd be included in the Feedback center? Well, no matter. I have a local copy, but before I post it: do those logs contain sensitive information? I don't know what's in them.

**Comment by ghost** [comment]
We don't object to posting traces. Don't post memory dumps?

If you grab a copy of tracefmt.exe you can have a gander at most of it. It contains logging about calls that failed/succeeded and so forth.

**Comment by JesseTG** [comment]
@pmartincic I don't have a memory dump, so I guess that's a good enough answer for me. Here you go!

[wsl.etl.zip](https://github.com/microsoft/WSL/files/8594863/wsl.etl.zip)

**Comment by ghost** [comment]
Registration for one of the relevant hyper-v components seems to be borked on your system.

**Reply by JesseTG** [reply]
Thanks for the tip. Which component, what's the problem, and what can I do about it? (I don't know how to analyze or make sense of the log I posted)

**Reply by ghost** [reply]
It pains me to say: I don't know if removing and re-adding the hyper-v feature would fix it. One of the files is missing, or the registration for it is gone?

**Reply by JesseTG** [reply]
> It pains me to say:

Why? Are you predicting a particularly painful fix?

> I don't know if removing and re-adding the hyper-v feature would fix it.

I'll disable all Hyper-V features, reboot, re-enable them, reboot, and then try again. I will also do this cycle with WSL. Can I do so with both features at once (for a total of two reboots), or should I do so in sequence (for a total of four reboots)?

> One of the files is missing, or the registration for it is gone?

I don't know, how can I tell?

**Reply by ghost** [reply]
This is where I need to depart the conversation :( I think it's related to vmsmb.dll

**Reply by JesseTG** [reply]
I appreciate your help, regardless. Can you at least point me to some related resource? I don't want to leave one of [*those*](https://xkcd.com/979/) threads behind.

**Reply by ghost** [reply]
Check what you have in Computer\HKEY_CLASSES_ROOT\CLSID\{FEA6E63B-45DA-4E32-8B3D-1873B2CE50E7}\InprocServer32
Do you have C:\Windows\System32\vmsmb.dll

**Reply by JesseTG** [reply]
I don't know exactly how registry paths work, so this is my best guess at what you're asking about:

```
PS C:\Users\Jesse> Get-Item 'HKLM:\SOFTWARE\Classes\CLSID\{FEA6E63B-45DA-4E32-8B3D-1873B2CE50E7}\InprocServer32'


    Hive: HKEY_LOCAL_MACHINE\SOFTWARE\Classes\CLSID\{FEA6E63B-45DA-4E32-8B3D-1873B2CE50E7}


Name                           Property
----                           --------
InprocServer32                 (default)      : C:\windows\System32\vmsmb.dll
                               ThreadingModel : Free
```

Does this look okay?

And yes, `C:\windows\System32\vmsmb.dll` does indeed exist.

**Reply by ghost** [reply]
Unfortunately that looks correct. This means that you're missing a load time link dependency of that module. If you have the know how you may be able to figure out what is the missing dependency. I can't vouch for it having only used its ancestor, but https://github.com/lucasg/Dependencies may help you with that task.

**Reply by JesseTG** [reply, marked as answer]
Too late, [I nuked the whole thing](https://superuser.com/questions/1718523/wsl-fails-to-start-in-windows-10-the-specified-module-could-not-be-found-er/1719260#1719260) and started over. Thanks anyway.

**Comment by slawkens** [comment]
I am running into exactly same issue. This happens with the WSL2 as backend.

My windows build is 19045.2546, so it should't be the issue.

I have enabled all required components and restarted PC afterwards. Hyper-V, Virtual Machine, Windows Subsystem.

The problem is only with WSL 2, because when I use:

wsl.exe --set-default-version 1

Then the installation works normally.

Any ideas? Can I provide any logs, if so, how?

Thanks for any guidance.

**Comment by Thaina** [comment]
same here. Need to use wsl 1 to install ubuntu

But then it said I can't use docker in wsl1

**Comment by Thaina** [comment]
I have use all the fix I can find, such as `wsl_update_x64.msi` and `MicrosoftProgram_Install_and_Uninstall.meta.diagcab`

Then I can finally use turn feature off to remove `Hyper-V` `Virtual Machine Platform` `Window Hypervisor Platform` and `Window Subsystem for Linux`, Restart, turn all of that on again, and then restart again

Now I can successfully install ubuntu

![image](https://github.com/user-attachments/assets/19a6d8cd-9ace-4185-a99e-0fe4d02cd58e)
